### PR TITLE
Automatically retry when rate limit is reached

### PR DESCRIPTION
**Pull request by budirec**

### CODE REVIEW

**Comment by SendGridDX** [comment]
[CLA assistant check](https://cla.sendgrid.com/sendgrid/php-http-client?pullRequest=23) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.sendgrid.com/sendgrid/php-http-client?pullRequest=23) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.sendgrid.com/check/sendgrid/php-http-client?pullRequest=23) it.</sub>

File: lib/Client.php [hunk 1]
```diff
@@ -134,10 +134,11 @@ private function buildUrl($queryParams = null)
       * @param string $url     the final url to call
       * @param array  $body    request body
       * @param array  $headers any additional request headers
+      * @param bool   $retryOnLimit should retry if rate limit is reach?
       *
       * @return Response object
       */
-    public function makeRequest($method, $url, $body = null, $headers = null)
+    public function makeRequest($method, $url, $body = null, $headers = null, $retryOnLimit = true)
```
Comment (thinkingserious) [inline lib/Client.php]: By default, this should be set to null to so we don't have a breaking change.

**Comment by budirec** [comment]
I don't believe handling other error is a good idea. it should be handled by the user.
Next improvement maybe to add flag to throw error instead of returning the error

**Comment by thinkingserious** [comment]
Hi @budirec,

Could you please update the tests? https://travis-ci.org/sendgrid/php-http-client/jobs/270245199#L292

Thanks!

File: lib/Client.php [hunk 2]
```diff
@@ -169,8 +175,17 @@ public function makeRequest($method, $url, $body = null, $headers = null)
         $responseHeaders = array_map('trim', $responseHeaders);
 
         curl_close($curl);
+     
+        $response = new Response($statusCode, $responseBody, $responseHeaders);
+
+        if ($statusCode == 429 && $retryOnLimit) {
+            $headers = $response->headers(true);
+            $sleepDurations = $headers['X-Ratelimit-Reset'] - time();
```
Comment (thinkingserious) [inline lib/Client.php]: Please explain the idea behind this algorithm.

I want to make sure we don't overload our servers when something has gone wrong. I was hoping for a backoff strategy that eventually fails after a certain period of time.

Comment (budirec) [reply lib/Client.php]: Every request we made to sendgrid will have X-Ratelimit-Reset in the header, where it tell us when the rate limit is reset in sendgrid.
So when ever we have error 429(rate limit reached) we use X-Ratelimit-Reset to determine how long the script need to wait before make another call to sendgrid.
This way we don't use a random time or predefined time.

**Comment by thinkingserious** [comment]
Hello @budirec,<br /><br/>Thanks again for the PR!<br />                                              <br />We want to show our appreciation by sending                                              you some swag. Could you please fill out                                              <a href="https://dx.sendgrid.com/swag">this form</a> so we can send it to you? Thanks!                                             <br /><br />Team SendGrid DX

**Pull request merged**